### PR TITLE
Notification Experiments 2 and 3

### PR DESCRIPTION
**Pull request by subsymbolic**
Task/Issue URL: https://app.asana.com/0/361428290920652/1112318511572574 & https://app.asana.com/0/361428290920652/1114095864120161
Tech Design URL: N/A

**Description**:
Add notification experiments 2 and 3

**Steps to test this PR**:

**TEST 1: Privacy Notification Trackers: Launch**
1. In `VariantManager` comment out everything other than the `me` variant and in `NotificationScheduler` change the time unit `val timeUnit = TimeUnit.DAYS` to minutes.
1. Run a clean version of the app
1. Browse to cnn.com
1. Kill the app and note that after a minute you see a notification informing of number of trackers blocked
1. Tap the notification text and enture the app launches
1. Ensure the "mnot_s_pp_8_0" and "mnot_l_pp_8_0" pixels have fired with numbers representing the trackers and upgrades
1. Relaunch the app and ensure the notification does not show again

**TEST 2: Privacy Notification Upgrades: Dismiss**
1. As above, in `VariantManager` comment out everything other than the `me` variant and in `NotificationScheduler` change the time unit `val timeUnit = TimeUnit.DAYS` to minutes.
1. Run a clean version of the app
1. Browse to facebook.com twice (once will be below threshold)
1. Kill the app and note that after a minute you see a notification informing of number of upgrades
1. Dismiss it
1. Ensure the "mnot_s_pp_0_2" and "mnot_c_pp_0_2" pixels have fired with numbers representing the number of trackers and upgrades
1. Relaunch the app and ensure the notification does not show again

**TEST 3: Privacy Notification No browsing**
1. In `VariantManager` comment out everything other than the `me` variant and in `NotificationScheduler` change the time unit `val timeUnit = TimeUnit.DAYS` to seconds.
1. Run a clean version of the app
1. Kill the app and note that after a moment you see a generic notification as no trackers or upgrades have happened yet

**TEST 4: Privacy Notification Browsing AND Upgrades**
1. In `VariantManager` comment out everything other than the `me` variant and in `NotificationScheduler` change the time unit `val timeUnit = TimeUnit.DAYS` to minutes.
1. Run a clean version of the app
1. Browse to cnn.com and facebook.com twice
1. Kill the app and note that after a minute you see a notification informing of number of trackers blocked and upgrades

**TEST 5: Privacy Clear Notification: Launch**
1. In `VariantManager` comment out everything other than `sc` and in `NotificationScheduler` change the time unit `val timeUnit = TimeUnit.DAYS` to seconds.
1. Run a clean version of the app
1. Background the app, note that after a few seconds the clear data notification shows
1. Tap it to launch the clear data screen
1. Ensure the "mnot_s_cd" and "mnot_l_cd" pixels have fired
1. Relaunch the app and ensure it does not show again

**TEST 6: Privacy Clear Notification: Dismiss**
1. As above, in `VariantManager` comment out everything other than `sc` and in `NotificationScheduler` change the time unit `val timeUnit = TimeUnit.DAYS` to seconds.
1. Run a clean version of the app
1. Background the app, note that after a few seconds the clear data notification shows
1. Dismiss it
1. Ensure the "mnot_s_cd" and "mnot_c_cd" pixels have fired
1. Relaunch the app and ensure it does not show again

**TEST 7: Migration**
1. Install the develop branch and run the app
1. Install this branch and ensure that the migration succeeds and nothing goes boom

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

### CODE REVIEW

**Comment by CDRussell** [comment]
For Test 2, i observed different values in the "launched" and "cancelled" pixels

Screenshot 2019-04-03 at 13 40 35

File: app/src/main/java/com/duckduckgo/app/notification/model/PrivacyProtectionNotification.kt [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    private val privacyProtectionCountDao: PrivacyProtectionCountDao
+) : NotificationSpec {
+
+    override val channel = NotificationRegistrar.ChannelType.TUTORIALS
+    override val systemId = NotificationRegistrar.NotificationId.PrivacyProtection
+    override val name = "Privacy protection"
+    override val icon = R.drawable.notification_sheild_lock
+    override val launchButton: String = context.getString(R.string.privacyProtectionNotificationLaunchButton)
+
+    private val trackers = runBlocking { privacyProtectionCountDao.getTrackersBlockedCount().toInt() }
```
Comment (CDRussell) [inline app/src/main/java/com/duckduckgo/app/notification/model/PrivacyProtectionNotification.kt]: I worry this disguises that this call is doing real work whenever it is created. It's not obvious when you instantiate this class it'll go retrieve from the DB. 

For example, i'm very surprised to see that this line invokes a DB call

`val specification = notification.specification`

File: app/src/main/java/com/duckduckgo/app/statistics/VariantManager.kt [hunk 2]
```diff
@@ -36,7 +40,15 @@ interface VariantManager {
...
             // a subgroup e.g a device API or specific language
-            Variant(key = "sc", weight = 1.0, features = emptyList())
+            Variant(key = "sc", weight = 1.0, features = emptyList()),
+
+            // Notification Content
+            Variant(key = "me", weight = 1.0, features = listOf(NotificationPrivacyDay1, NotificationSuppressClearDataDay3)),
+            Variant(key = "mi", weight = 1.0, features = listOf(NotificationClearDataDay1, NotificationSuppressClearDataDay3)),
+
+            // Notification Add New
+            Variant(key = "mf", weight = 1.0, features = listOf()),
```
Comment (CDRussell) [inline app/src/main/java/com/duckduckgo/app/statistics/VariantManager.kt]: I assume `mf` is needed but i can't immediately see why `sc` wouldn't suffice?

File: app/src/main/java/com/duckduckgo/app/browser/BrowserWebViewClient.kt [hunk 3]
```diff
@@ -134,7 +136,7 @@ class BrowserWebViewClient @Inject constructor(
     @WorkerThread
     override fun shouldInterceptRequest(webView: WebView, request: WebResourceRequest): WebResourceResponse? {
         Timber.v("Intercepting resource ${request.url} on page $currentUrl")
-        return requestInterceptor.shouldIntercept(request, webView, currentUrl, webViewClientListener)
+        return requestInterceptor.shouldIntercept(request, webView, currentUrl, webViewClientListener, privacyProtectionCountDao)
```
[the remark this answers is not in the document]
Comment (subsymbolic) [reply app/src/main/java/com/duckduckgo/app/browser/BrowserWebViewClient.kt]: Good point! I'll inject at the next layer.

File: app/src/main/java/com/duckduckgo/app/notification/NotificationScheduler.kt [hunk 4]
```diff
@@ -21,117 +21,104 @@ import android.app.PendingIntent.getService
...
-            "com.duckduckgo.privacytips.autoclear",
-            NotificationRegistrar.ChannelType.PRIVACY_TIPS,
-            "Update auto clear data",
-            R.drawable.notification_fire,
-            R.string.clearNotificationTitle,
-            R.string.clearNotificationDescription
-        )
-    }
-
     fun scheduleNextNotification(scope: CoroutineScope = GlobalScope) {
```
[the remark this answers is not in the document]
Comment (subsymbolic) [reply app/src/main/java/com/duckduckgo/app/notification/NotificationScheduler.kt]: Good suggestion, I've made the change.

File: app/src/main/java/com/duckduckgo/app/notification/NotificationScheduler.kt [hunk 5]
```diff
@@ -21,117 +21,104 @@ import android.app.PendingIntent.getService
...
+        lateinit var notification: SchedulableNotification
         lateinit var notificationDao: NotificationDao
-        lateinit var settingsDataStore: SettingsDataStore
         lateinit var pixel: Pixel
 
         override fun doWork(): Result {
 
-            if (settingsDataStore.automaticallyClearWhatOption != ClearWhatOption.CLEAR_NONE) {
-                Timber.v("No need for notification, user already has clear option set")
+            val canShow = runBlocking { notification.canShow() }
```
[the remark this answers is not in the document]
Comment (subsymbolic) [reply app/src/main/java/com/duckduckgo/app/notification/NotificationScheduler.kt]: I've been using `runBlocking` when not on the main thread but happy to make it a team style to only use when absolutely necessary. `CorouteineWorker` is a great idea 👍

File: app/src/main/java/com/duckduckgo/app/statistics/VariantManager.kt (hunk 2 above)
Comment (subsymbolic) [reply app/src/main/java/com/duckduckgo/app/statistics/VariantManager.kt]: Good shout, it should suppress the clear data notification in prod, I've updated it.

File: app/src/main/java/com/duckduckgo/app/notification/model/PrivacyProtectionNotification.kt (hunk 1 above)
Comment (subsymbolic) [reply app/src/main/java/com/duckduckgo/app/notification/model/PrivacyProtectionNotification.kt]: I've turned it into a suspend function called `buildSpecification` to make the fact that it's doing heavy lifting more apparent.

[the remark this answers is not in the document]
Comment (subsymbolic) [reply app/src/main/java/com/duckduckgo/app/notification/model/PrivacyProtectionNotification.kt]: This is now gone (see previous runBlocking comments)

**Comment by subsymbolic** [comment]
Thanks for the review @CDRussell, coroutine code is much improved after your input. Over to you for rereview. Fyi I could never reproduce the different launch and cancel pixel issue however had a hunch it was the `pixelSuffix` variable in the `NotificationHandlerService` so updated that. Can you give it a good test?

File: app/src/main/java/com/duckduckgo/app/notification/model/PrivacyProtectionNotification.kt (hunk 1 above)
Comment (CDRussell) [reply app/src/main/java/com/duckduckgo/app/notification/model/PrivacyProtectionNotification.kt]: Nice change; I think that's much clearer 👌

**Comment by CDRussell** [comment]
Code changes look great, but i'm still seeing the wrong `pixelSuffix`. I can replicate it fairly easily but i'm still not sure what's going on with it. ping me if you want to pair on it?

**Comment by CDRussell** [comment]
After some thorough testing with @subsymbolic, we've concluded that the discrepancy observed above with the wrong `pixelSuffix` is likely caused by Android system bug.

LGTM.

**Details for posterity**
The bug only manifests itself when clearing the app data and re-trying; it doesn't happen after a fresh install.

Somehow, Android is delivering an old `Intent` which is used when you swipe the notification away under the developer-driven process of clearing data repeatedly but not uninstalling.

To confirm this theory, we added a `UUID` to the `Intent`, and observed it was the expected `UUID` that was retrieved from the `Intent` the first time around after a clean install. However, after clearing data and trying again, the same `UUID` as above was returned instead of the new one which should have been delivered.

**Pull request merged**